File: chazm-model/src/main/java/runtimemodels/chazm/model/event/AgentEventFactory.java
package runtimemodels.chazm.model.event;

import runtimemodels.chazm.api.entity.Agent;

/**
 * The {@linkplain AgentEventFactory} interface defines the API for constructing {@linkplain AgentEvent}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface AgentEventFactory {

    /**
     * Constructs an {@linkplain AgentEvent}.
     *
     * @param category the {@linkplain EventType}.
     * @param agent    the {@linkplain Agent}.
     * @return a {@linkplain AgentEvent}.
     */
    AgentEvent build(EventType category, Agent agent);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/AgentEventFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Agent
import runtimemodels.chazm.model.event.EventType

/**
 * The [AgentEventFactory] interface defines the API for constructing [AgentEvent]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface AgentEventFactory {
    /**
     * Constructs an [AgentEvent].
     *
     * @param category the [EventType].
     * @param agent    the [Agent].
     * @return a [AgentEvent].
     */
    fun build(category: EventType, agent: Agent): AgentEvent
}
